synapse/storage/controllers/purge_events.py

```diff
@@ -203,3 +212,54 @@ async def _delete_state_groups(
             room_id,
             groups_to_sequences,
         )
+
+    @wrap_as_background_process("_clear_unreferenced_state_groups_loop")
+    async def _clear_unreferenced_state_groups_loop(self) -> None:
+        """Background task that deletes any state groups that may be pending
+        deletion."""
+
+        # Look for state groups that can be cleaned up.
+        search_limit = 100
+        next_set = await self.stores.main.get_state_groups(
+            self._last_checked_state_group + 1, search_limit
+        )
+        if len(next_set) < search_limit:
+            self._last_checked_state_group = 0
+        else:
+            self._last_checked_state_group = list(next_set)[-1]
```

Comment:
```suggestion
            self._last_checked_state_group = max(next_set)
```

I don't think we can assume sets are ordered in a sane way, so `list(next_set)[-1]` will potentially return a random entry.